apiserver/facades/client/resources/facade.go

```diff
@@ -26,82 +26,123 @@ var logger = loggo.GetLogger("juju.apiserver.resources")
 // Backend is the functionality of Juju's state needed for the resources API.
 type Backend interface {
 	// ListResources returns the resources for the given application.
-	ListResources(service string) (resource.ApplicationResources, error)
+	ListResources(application string) (resource.ApplicationResources, error)
 
-	// AddPendingResource adds the resource to the data store in a
+	// AddPendingResource adds the resource to the data backend in a
 	// "pending" state. It will stay pending (and unavailable) until
 	// it is resolved. The returned ID is used to identify the pending
 	// resources when resolving it.
 	AddPendingResource(applicationID, userID string, chRes charmresource.Resource) (string, error)
 }
 
-// CharmStore exposes the functionality of the charm store as needed here.
-type CharmStore interface {
-	// ListResources composes, for each of the identified charms, the
-	// list of details for each of the charm's resources. Those details
-	// are those associated with the specific charm revision. They
-	// include the resource's metadata and revision.
-	ListResources([]charmstore.CharmID) ([][]charmresource.Resource, error)
+// API is the public API facade for resources.
+type API struct {
+	// backend is the data source for the facade.
+	backend Backend
 
-	// ResourceInfo returns the metadata for the given resource.
-	ResourceInfo(charmstore.ResourceRequest) (charmresource.Resource, error)
+	factory func(chID CharmID) (NewCharmRepository, error)
 }
 
-// Facade is the public API facade for resources.
-type Facade struct {
-	// store is the data source for the facade.
-	store Backend
-
-	newCharmstoreClient func() (CharmStore, error)
+type APIv1 struct {
+	*API
 }
 
-// NewPublicFacade creates a public API facade for resources. It is
+// NewFacadeV2 creates a public API facade for resources. It is
 // used for API registration.
-func NewPublicFacade(st *state.State, _ facade.Resources, authorizer facade.Authorizer) (*Facade, error) {
+func NewFacadeV2(ctx facade.Context) (*API, error) {
+	authorizer := ctx.Auth()
 	if !authorizer.AuthClient() {
 		return nil, apiservererrors.ErrPerm
 	}
 
+	st := ctx.State()
 	rst, err := st.Resources()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+
 	controllerCfg, err := st.ControllerConfig()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	newClient := func() (CharmStore, error) {
-		return charmstore.NewCachingClient(state.MacaroonCache{st}, controllerCfg.CharmStoreURL())
+
+	m, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	modelCfg, err := m.Config()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	factory := func(chID CharmID) (NewCharmRepository, error) {
+		switch chID.URL.Schema {
+		case "ch":
```

Comment:
Going to mega annoying and ask that we use `charm.CharmHub.Matches`.

Reply:
done